### PR TITLE
Record dimension coordinate cell characteristics

### LINKED ISSUE #692: Record cell characteristics from aggregation cell conditions
Since 3.15.2, the `cells` keyword of `cf.aggregate` allows fields to be aggregated sperately according to the characteristics of their dimension coordinates - namely cell spacing and  cell size.

It would be useful to automatically record these characteristics on the aggregated fields, so that it is easy to see afterwards which conditions were passed.

Since the characteristics apply to dimension coordinate constructs only, they should be recorded on those constructs.

I propose four new `DimensionCoordinate` methods of `[del|get|has|set]_cell_characterstics` to allow the cell characteristics to be stored and subsequently accessed.

### PR DESCRIPTION
**Pull request by davidhassell**
Fixes #692

### CODE REVIEW

File: Changelog.rst [hunk 1]
```diff
@@ -3,6 +3,8 @@ version 3.15.4
 
 **2023-??-??**
 
+* Record dimension coordinate cell characteristics
+  (https://github.com/NCAS-CMS/cf-python/issues/692)
```
Comment (sadielbartholomew) [inline Changelog.rst:7]: Might be useful to explicitly mention the new methods, something like:

```suggestion
  (https://github.com/NCAS-CMS/cf-python/issues/692)
* New set of methods to query, set, del, and `get_cell_characterstics`
```
?

File: cf/dimensioncoordinate.py [hunk 2]
```diff
@@ -710,6 +765,41 @@ def create_bounds(
 
         return bounds
 
+    def del_cell_characteristics(self, default=ValueError()):
+        """Remove the cell characteristics.
+
+        A cell characterisitic is assumed to valid for each cell. Cell
```
Comment (sadielbartholomew) [inline cf/dimensioncoordinate.py:771]:
```suggestion
        A cell characteristic is assumed to be valid for each cell. Cell
```

File: cf/dimensioncoordinate.py [hunk 3]
```diff
@@ -776,66 +866,80 @@ def get_bounds(self, default=ValueError(), **kwargs):
...
-    #                print(4)
-    #            return False
-    #
-    #        self.period(period)
-    #
-    #        return True
+    def get_cell_characteristics(self, default=ValueError()):
+        """Return cell characteristics.
+
+        A cell characterisitic is assumed to valid for each cell. Cell
```
Comment (sadielbartholomew) [inline cf/dimensioncoordinate.py:872]:
```suggestion
        A cell characteristic is assumed to be valid for each cell. Cell
```

File: cf/dimensioncoordinate.py [hunk 4]
```diff
@@ -776,66 +866,80 @@ def get_bounds(self, default=ValueError(), **kwargs):
...
+
+        """
+        out = self._get_component("cell_characteristics", default=None)
+        if out is None:
+            if default is None:
+                return
+
+            return self._default(
+                default,
+                f"{self.__class__.__name__} has no 'cell_characteristics "
```
Comment (sadielbartholomew) [inline cf/dimensioncoordinate.py:912]:
```suggestion
                f"{self.__class__.__name__} has no 'cell_characteristics' "
```

File: cf/dimensioncoordinate.py [hunk 5]
```diff
@@ -776,66 +866,80 @@ def get_bounds(self, default=ValueError(), **kwargs):
...
+            )
+
+        from copy import deepcopy
+
+        return deepcopy(out)
+
+    def has_cell_characteristics(self):
+        """Whether or not there are any cell characteristics.
+
+        A cell characterisitic is assumed to valid for each cell. Cell
```
Comment (sadielbartholomew) [inline cf/dimensioncoordinate.py:923]:
```suggestion
        A cell characteristic is assumed to be valid for each cell. Cell
```

File: cf/dimensioncoordinate.py [hunk 6]
```diff
@@ -908,6 +1011,164 @@ def roll(self, axis, shift, inplace=False, i=False):
...
+
+        .. seealso: `del_bounds`, `get_bounds`, `has_bounds`, `set_data`
+
+        :Parameters:
+
+            bounds: `Bounds`
+                The bounds to be inserted.
+
+            copy: `bool`, optional
+                If False then do not copy the bounds prior to
```
Comment (sadielbartholomew) [inline cf/dimensioncoordinate.py:1027]: To be clearer by avoiding a double negative:

```suggestion
                If True then copy the bounds prior to
```

File: cf/dimensioncoordinate.py [hunk 7]
```diff
@@ -908,6 +1011,164 @@ def roll(self, axis, shift, inplace=False, i=False):
...
+        None
+
+        """
+        self._del_component("cell_characteristics", default=None)
+        super().set_bounds(bounds, copy=copy)
+
+    def set_cell_characteristics(self, cellsize, spacing):
+        """Set cell characteristics.
+
+        A cell characterisitic is assumed to valid for each cell. Cell
```
Comment (sadielbartholomew) [inline cf/dimensioncoordinate.py:1060]:
```suggestion
        A cell characteristic is assumed to be valid for each cell. Cell
```

File: cf/dimensioncoordinate.py [hunk 8]
```diff
@@ -908,6 +1011,164 @@ def roll(self, axis, shift, inplace=False, i=False):
...
+
+        :Parameters:
+
+            data: `Data`
+                The data to be inserted.
+
+                {{data_like}}
+
+            copy: `bool`, optional
+                If False then do not copy the data prior to
```
Comment (sadielbartholomew) [inline cf/dimensioncoordinate.py:1132]: Similar to a previous suggestion:

```suggestion
                If True then copy the data prior to
```

**Review by sadielbartholomew — APPROVED** [review]
Nice idea. Implementation is elegant and testing is thorough. I've made some minor suggestions, mostly to fix repeat/copy-and-pasted typos, but this is all good overall so please merge when ready. Thanks.

File: cf/aggregate.py [hunk 9]
```diff
@@ -2358,18 +2358,26 @@ def aggregate(
 
             >>> x = cf.aggregate(fl, cells=cf.climatology_cells())
 
+            **Storage of conditions**
+
+            All returned field or domain constructs that have passed
+            dimension coordinate cell conditions will have those
+            conditions stored on the approriate dimension coordinate
```
Comment (sadielbartholomew) [inline cf/aggregate.py:2365]: One more typo fix, after I thought to check on any other repeats:

```suggestion
            conditions stored on the appropriate dimension coordinate
```

File: cf/aggregate.py [hunk 10]
```diff
@@ -3058,6 +3070,52 @@ def aggregate(
...
+            are updated in-place.
+
+    :Returns:
+
+        `None`
+
+    """
+    for m in output_meta:
+        for value in m.axis.values():
+            dim_index = value["dim_coord_index"]
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/aggregate.py:3095]: Hi Sadie - we're OK here, as the key `"dim_coord_index"` is guaranteed to exist. If for some reason it isn't there then that is a bug which needs fixing, so there is no sensible default value. Thanks.

**Pull request merged**